src/config/mod.rs

```diff
@@ -0,0 +1,195 @@
+}
+
+fn start_ws(config: &YamlConfig, access_token: String) -> Result<WsLongLivedAccessTokenResponse> {
+    let ws_url = format!("ws://{}/api/websocket", config.ha.host);
+    let (mut socket, _) = connect(Url::parse(ws_url.as_str()).unwrap()).expect("Can't connect");
+
+    socket.read_message().expect("Error reading message");
+    let req = WsAuthRequest {
+        auth_type: "auth".to_string(),
+        access_token,
+    };
+
+    let req_as_str = Message::Text(serde_json::to_string(&req)?);
+    socket.write_message(req_as_str).unwrap();
+
+    socket.read_message().expect("Error reading message");
+
+    let req2 = WsLongLivedAccessTokenRequest {
+        id: 11,
+        command_type: "auth/long_lived_access_token".to_string(),
+        client_name: "Halcyon".to_string(),
+        lifespan: 365,
+    };
+
+    let req2_as_str = Message::Text(serde_json::to_string(&req2)?);
+    socket.write_message(req2_as_str).unwrap();
+
+    let msg3 = socket.read_message().expect("Error reading message");
+
+    let response: WsLongLivedAccessTokenResponse =
+        serde_json::from_str(msg3.into_text().unwrap().as_str())?;
+
+    if response.success {
+        Ok(response)
+    } else {
+        Err("Could not get long lived token from websocket (perhaps one already exists for halcyon in home assistant)".into())
+    }
+}
+
+impl YamlConfig {
+    pub fn update_device_id_if_needed(self, file_name: &str) -> Result<Self> {
+        let new_config = match self.ha.device_id {
+            None => {
+                let ha_config = HaConfig {
+                    device_id: Some(Uuid::new_v4().to_string()),
+                    ..self.ha
+                };
+                let config = YamlConfig { ha: ha_config };
+                let new_config_str = serde_yaml::to_string(&config)?;
+                let mut f = std::fs::OpenOptions::new()
+                    .write(true)
+                    .truncate(true)
+                    .open(file_name)?;
+                f.write_all(new_config_str.as_bytes())?;
+                println!("no device-id found in config, so we are making one for you");
+                config
+            }
+            Some(_) => self,
+        };
+        Ok(new_config)
+    }
+
+    pub async fn update_long_lived_access_token_if_needed(self, file_name: &str) -> Result<Self> {
+        let new_config = match self.ha.long_lived_token {
+            None => {
+                let access_token_resp = wait_for_token(&self).await?;
+                let long_lived_access_token_resp = start_ws(&self, access_token_resp.access_token)?;
+                let ha_config = HaConfig {
+                    long_lived_token: long_lived_access_token_resp.result,
+                    ..self.ha
+                };
+                let config = YamlConfig { ha: ha_config };
+                let new_config_str = serde_yaml::to_string(&config)?;
+                let mut f = std::fs::OpenOptions::new()
```

Comment:
Might be nice to extract this file updating into a new separate method, so it can be reused?